### PR TITLE
Add benchmarking tests for Parser.

### PR DESCRIPTION
**Pull request by markcol**
Add a benchmarking test for the parser to detect performance
regressions during development. The Criterion benchmark crate is
used to perform the benchmarks and provides a number of useful
metrics for evaluating changes between runs.

To run the benchmarks, run the following command:

  cargo bench -- --test [--nocapture]

This has been added to the .travis.yml file so that performance
benchmarks are run as part of the standard suite of continuous
integration tests.

In order to preserve benchmarking comparisons between releases, the
.criterion directory is checked into git, but is also in .gitignore.
This allows stable performance comparisons between releases and
during CI runs, rather than being overwritten with each checkin.

A new version of the .criterion directory should be checked into git
manually at each release to capture the new performance baseline
for future development.

### CODE REVIEW

**Comment by LeopoldArkham** [comment]
Thanks, I've wanted this for a while.
Three things:
- It looks like Criterion can only be run on nightly.
- I think using `full.toml` as a benchmark might yield more meaningful results since it contains one of everything. `indented.toml` only contains 3 KV pairs and might miss a potentially slow codepath entirely.
Alternatively, we can look at what other TOML parsers are using for benchmarking and use that.
- With regards to the criterion directory, you're saying its latest local version should be manually committed every time Molten's version changes? I've never seen that done so I want to make sure I understand you.

File: benches/parser_benchmark.rs [hunk 1]
```diff
@@ -0,0 +1,29 @@
+//! Parser benchmark tests to identify any performance regressions
+//! in the parser.
+
+// For details on using Criterion, see the documentation at
+// https://japaric.github.io/criterion.rs/book/index.html
+extern crate criterion;
+#[macro_use()]
```
Comment (LeopoldArkham) [inline benches/parser_benchmark.rs]: I don't think this line does anything

**Comment by markcol** [comment]
@LeopoldArkham to address your points:

1. Yes, Criterion can only be run on nightly. The default benchmark tools also only run on nightly (they require "#![feature(test)]", so II didn't think that would be a problem. We could only enable this and run it against nightly builds, or we could defer it until later when benchmarking is pulled up to stable. Your call.

1. Agree. I have changed the input to use full.toml. We can always change it later if we want something more exhaustive, etc. Looking at other parser benchmarks would also be a good idea.

1. Yes, that is correct. The baseline data is stored in the .criteria directory so that Criterion can evaluate performance against the last run. The basic idea is to check in a new baseline only when we release so that comparisons are done against the last released build. We can automate the check in on release as part of the release push via Travis, or Cargo magic.

**Comment by markcol** [comment]
Now the build, test, and bench paths all work with stable, beta and nightly builds. To get benchmarks  to use Criterion, you need to run them with the following command:

   `cargo bench --features nightly -- --tests --nocapture`

Otherwise the Criterion crate will not be included as a dependency, and the benchmark code won't be compiled. This is handled properly in the `.travis.yml` file so that benchmarks are only run on nightly builds, but tests are run on stable and beta, as well as nightly.

If you run `cargo bench` without the `--features nightly` it will run successfully, reporting that there are zero tests to run.

**Comment by LeopoldArkham** [comment]
Great, thanks!

**Pull request merged**